oadp-operator/controllers/restic.go

```diff
@@ -1,14 +1,321 @@
+										FieldRef: &v1.ObjectFieldSelector{
+											FieldPath: "/credentials/cloud",
+										},
+									},
+								},
+								{
+									Name: "GOOGLE_APPLICATION_CREDENTIALS",
+									ValueFrom: &v1.EnvVarSource{
+										FieldRef: &v1.ObjectFieldSelector{
+											FieldPath: "/credentials-gcp/cloud",
+										},
+									},
+								},
+								{
+									Name: "AZURE_CREDENTIALS_FILE",
+									ValueFrom: &v1.EnvVarSource{
+										FieldRef: &v1.ObjectFieldSelector{
+											FieldPath: "/credentials-azure/cloud",
+										},
+									},
+								},
+								{
+									Name: "VELERO_SCRATCH_DIR",
+									ValueFrom: &v1.EnvVarSource{
+										FieldRef: &v1.ObjectFieldSelector{
+											FieldPath: "/scratch",
+										},
+									},
+								},
+							},
+						},
+					},
+					InitContainers: []v1.Container{
+						{
+							Name: "setup-certificate-secret",
+							Command: []string{
+								"sh",
+								"-ec",
+								">-",
+								"cp /etc/ssl/certs/* /certs/; ln -sf /credentials/ca_bundle.pem",
+								"/certs/ca_bundle.pem;",
+							},
+							Resources:                v1.ResourceRequirements{},
+							TerminationMessagePath:   "/dev/termination-log",
+							TerminationMessagePolicy: v1.TerminationMessagePolicy("File"),
+							VolumeMounts: []v1.VolumeMount{
+								{
+									Name:      "certs",
+									MountPath: "/certs",
+								},
+								{
+									Name:      string(oadpv1alpha1.DefaultPluginAWS),
+									MountPath: "/credentials",
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	if velero.Spec.EnableRestic != nil && !*velero.Spec.EnableRestic {
+		// If velero Spec enableRestic exists and is false, attempt to delete.
+		r.Delete(r.Context, &ds) //TODO: delete fail logic?
+		return true, nil
+	}
+	// Dynamically add to daemonset definition
+	// If the default velero plugins contains cloud provider, attach VolumeSource
+	for provider, providerSecretName := range cloudProviderSecretNames {
+		if contains(provider, velero.Spec.DefaultVeleroPlugins) {
+			ds.Spec.Template.Spec.Volumes = append(
+				ds.Spec.Template.Spec.Volumes,
+				v1.Volume{
+					Name: string(provider),
+					VolumeSource: v1.VolumeSource{
+						Secret: &v1.SecretVolumeSource{
+							SecretName: providerSecretName,
+						},
+					},
+				},
+			)
+		}
+	}
+
+	// Check if Daemonset already exists
+	//make a copy of ds to be used in r.Get
+	existingDS := *ds.DeepCopy()
+	if err := r.Get(r.Context, r.NamespacedName, &existingDS); err != nil {
+		if errors.IsNotFound(err) { // Daemonset not found so create Daemonset
+			if err := r.Create(r.Context, &ds); err != nil {
+				return false, err
+			}
+		}
+	} else {
+		// Daemonset found, check if equal.
+		if err := r.Update(r.Context, &ds); err != nil { // Update daemonset
+			return false, err
+		}
+	}
+
 	return true, nil
 }
+
+func getVeleroResourceReqs(velero *oadpv1alpha1.Velero) v1.ResourceRequirements {
```

Comment:
We will need to handle the merge here as I think this will collide with @shubham-pampattiwar PR as well